File: src/WebUI/BoardGameAngular/ClientApp/src/app/config/config.service.ts
import { Injectable, Inject } from '@angular/core';

class BaseUrl {
  constructor(protected readonly baseUrl: string) { }
}

@Injectable()
export class UrlConfigService extends BaseUrl{
  public readonly userBackendUrl: UserUrl
  public readonly roomBackendUrl: RoomUrl
  public readonly gameBackendUrl: GameUrl
  public readonly bigTwoBackendUrl: BigTwoUrl
  
  constructor(@Inject('BASE_URL') baseUrl: string) {
    super(`${baseUrl}api`);
    this.userBackendUrl = new UserUrl(`${this.baseUrl}/User`);
    this.roomBackendUrl = new RoomUrl(`${this.baseUrl}/Room`);
    this.gameBackendUrl = new RoomUrl(`${this.baseUrl}/Game`);
    this.bigTwoBackendUrl = new BigTwoUrl(`${this.baseUrl}/BigTwo`);
  }
}

export class UserUrl extends BaseUrl{
  constructor(baseUrl: string) {
    super(baseUrl);
  }
  
  get UserStatus(): string {
    return `${this.baseUrl}/Status`;
  }
  get Login(): string {
    return `${this.baseUrl}/Login`;
  }
  get Register(): string {
    return `${this.baseUrl}/RegisterAndLogin`;
  }
  get Update(): string {
    return `${this.baseUrl}/update`;
  }
  get Info(): string {
    return `${this.baseUrl}`;
  }
}

export class RoomUrl extends BaseUrl {
  constructor(baseUrl: string) {
    super(baseUrl);
  }

  get Start(): string {
    return `${this.baseUrl}/Start`;
  }
  get List(): string {
    return `${this.baseUrl}`;
  }
  get Create(): string {
    return `${this.baseUrl}`;
  }
  get Join(): string {
    return `${this.baseUrl}`;
  }
  get Leave(): string {
    return `${this.baseUrl}`;
  }
}

export class GameUrl extends BaseUrl {
  constructor(baseUrl: string) {
    super(baseUrl);
  }

  get List(): string {
    return `${this.baseUrl}`;
  }
}

export class BigTwoUrl extends BaseUrl {
  constructor(baseUrl: string) {
    super(baseUrl);
  }

  get HandCards(): string {
    return `${this.baseUrl}/HandCards`;
  }
  get SelectCards(): string {
    return `${this.baseUrl}/SelectCards`;
  }
  get PlayCards(): string {
    return `${this.baseUrl}/PlayCards`;
  }
}

File: src/WebUI/BoardGameAngular/Services/Config/ConfigService.cs
namespace BoardGameAngular.Services.Config
{
    public class ConfigService
    {
        public string UserLogin => _userBase + "/Login";
        public string UserRegister => _userBase + "/Register";
        public string UserUpdate => _userBase + "/Update";
        private string _userBase => _baseUrl + "/User";

        public string UserStatus => _roomBase + "/User";
        public string RoomStart => _roomBase + "/Start";
        public string LeaveRoom => _roomBase;
        public string JoinRoom => _roomBase;
        public string CreateRoom => _roomBase;
        public string ListRoom => _roomBase;
        private string _roomBase => _baseUrl + "/Room";

        public string ListGame => _gameBase;
        public string StartGame => _gameBase + "/StartGame";
        private string _gameBase => _baseUrl + "/Game";

        public string GameStatus => _bigTwoBase + "/GameStatus";
        public string PlayCards => _bigTwoBase + "/PlayCards";
        public string SelectCards => _bigTwoBase + "/SelectCards";
        public string HandCards => _bigTwoBase + "/HandCards";
        private string _bigTwoBase => _boardgameBase + "/BigTwo";

        private string _boardgameBase => _baseUrl + "/boardgame";

        private readonly string _baseUrl;

        public ConfigService(string backendDomain)
        {
            _baseUrl = backendDomain + "/api";
        }
    }
}
